src/request/url.rs

```diff
@@ -70,70 +67,47 @@ impl Url {
     /// Create a `Url` from a `rust-url` `Url`.
     pub fn from_generic_url(raw_url: url::Url) -> Result<Url, String> {
         // Create an Iron URL by extracting the special scheme data.
-        match raw_url.scheme_data {
-            SchemeData::Relative(data) => {
+        match raw_url.cannot_be_a_base() {
+            false => {
                 // Extract the port as a 16-bit unsigned integer.
-                let port: u16 = match data.port {
-                    // If explicitly defined, unwrap it.
+                let port: u16 = match raw_url.port_or_known_default() {
+                    // If explicitly defined or has a known default, unwrap it.
                     Some(port) => port,
 
                     // Otherwise, use the scheme's default port.
-                    None => {
-                        match whatwg_scheme_type_mapper(&raw_url.scheme) {
-                            SchemeType::Relative(port) => port,
-                            _ => return Err(format!("Invalid special scheme: `{}`",
-                                                    raw_url.scheme))
-                        }
-                    }
+                    None => return Err(format!("Invalid special scheme: `{}`",
+                                                    raw_url.scheme())),
                 };
-
                 // Map empty usernames to None.
-                let username = match &*data.username {
+                let username = match raw_url.username() {
                     "" => None,
-                    _ => Some(data.username)
+                    _ => Some(raw_url.username())
                 };
-
                 // Map empty passwords to None.
-                let password = match data.password {
+                let password = match raw_url.password() {
                     None => None,
                     Some(ref x) if x.is_empty() => None,
                     Some(password) => Some(password)
                 };
-
                 Ok(Url {
-                    scheme: raw_url.scheme,
-                    host: data.host,
+                    scheme: raw_url.scheme().to_string(),
+                    // `unwrap` is safe here because urls that cannot be a base don't have a host
+                    host: raw_url.host().unwrap().to_owned(),
                     port: port,
-                    path: data.path,
-                    username: username,
-                    password: password,
-                    query: raw_url.query,
-                    fragment: raw_url.fragment
+                    path: raw_url.path_segments().into_iter().flat_map(|x| x).map(|x| x.to_string()).collect(),
+                    username: username.map(|s| s.to_string()),
+                    password: password.map(|s| s.to_string()),
+                    query: raw_url.query().map(|s| s.to_string()),
+                    fragment: raw_url.fragment().map(|s| s.to_string()),
```

Comment:
Though I am unsure if that's a desirable change or purely subject to
bikeshedding.

On Tue, May 03, 2016 at 12:44:39PM -0700, Carol (Nichols || Goulding) wrote:
> >                      port: port,
> > -                    path: data.path,
> > -                    username: username,
> > -                    password: password,
> > -                    query: raw_url.query,
> > -                    fragment: raw_url.fragment
> > +                    path: raw_url.path_segments().into_iter().flat_map(|x| x).map(|x| x.to_string()).collect(),
> > +                    username: username.map(|s| s.to_string()),
> > +                    password: password.map(|s| s.to_string()),
> > +                    query: raw_url.query().map(|s| s.to_string()),
> > +                    fragment: raw_url.fragment().map(|s| s.to_string()),
> 
> Sorry, do you mean make this change?
> 
> ```diff
> - fragment: raw_url.fragment().map(|s| s.to_string()),
> + fragment: raw_url.fragment().to_string(),
> ```
> 
> This doesn't work because `url::Url::fragment` returns `Option<&str>`, and `Option<&str>` doesn't implement `Display`. Or is there another way to use `std::string::ToString::to_string`?
> 
> ---
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/iron/iron/pull/452/files/bca05d75107e7bb3896064e78633334d663d5045#r61943140